src/Testcontainers/Clients/DockerContainerOperations.cs

```diff
@@ -208,11 +211,41 @@ public async Task<string> RunAsync(IContainerConfiguration configuration, Cancel
         }
       }
 
+      if (configuration.Reuse.HasValue && configuration.Reuse.Value)
+      {
+        var hash = CreateReuseHash(createParameters);
+        createParameters.Labels.Add(TestcontainersClient.TestcontainersReuseHashLabel, hash);
+
+        var containers = await Docker.Containers.ListContainersAsync(new ContainersListParameters()
+        {
+          Filters = new FilterByProperty {
+              { "label", $"{TestcontainersClient.TestcontainersReuseHashLabel}={hash}" }
+            },
+          Limit = 1,
+        }, ct)
+          .ConfigureAwait(false);
+
+        var container = containers.FirstOrDefault();
+        if (container != default) {
+          return container.ID;
+        }
+      }
```

Comment:
It feels like there are better locations for this. But since this is the only location where we can access the `CreateContainerParams` used to create the container, I put the reuse logic here.

Reply:
Maybe this should belong to `DockerContainer.UnsafeCreateAsync()`?

Reply:
Interesting idea. I have not thought about using the `createParameters` instance. I was thinking about extending the `ResourceConfiguration` class and its implementations. I believe this has several advantages:

1. We need to calculate the hash value for the other resources (like network and volume) too. Each resource configuration derives from this class.
2. It is simpler to test. We can expose generating the hash value for each resource configuration through the public abstract `Resource` class (the resource interfaces). Therefore we do not need to create the resource at all.

    ```csharp
    _ = new ContainerBuilder().Build().GetHash();
    _ = new VolumeBuilder().Build().GetHash();
    ```
3. This should cover all resources at once. We probably "just" need to ignore individual properties of each resource from being considered in calculating the hash value if they change (like the session ID).